lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,773 @@
+		fn $transition(self, msg: $msg) -> StateMachine {
+			match self {
+				$(
+					Self::$from_state(s) => match s.transition(msg) {
+						Ok(new_state) => StateMachine::$to_state(new_state),
+						Err(abort_reason) => StateMachine::NegotiationAborted(NegotiationAborted(abort_reason)),
+					}
+				 )*
+				_ => StateMachine::NegotiationAborted(NegotiationAborted(AbortReason::UnexpectedCounterpartyMessage)),
+			}
+		}
+	};
+	(LOCAL_OR_REMOTE_CHANGE, $to_local_transition: ident, $to_remote_transition: ident, $msg: ty) => {
+		define_state_machine_transitions!($to_local_transition, $msg, [
+			FROM RemoteChange, TO LocalChange,
+			FROM RemoteTxComplete, TO LocalChange
+		]);
+		define_state_machine_transitions!($to_remote_transition, $msg, [
+			FROM LocalChange, TO RemoteChange,
+			FROM LocalTxComplete, TO RemoteChange
+		]);
+	};
+}
+
+impl StateMachine {
+	fn new(feerate_sat_per_kw: u32, is_initiator: bool, tx_locktime: PackedLockTime) -> Self {
+		let context = NegotiationContext {
+			tx_locktime,
+			holder_is_initiator: is_initiator,
+			received_tx_add_input_count: 0,
+			received_tx_add_output_count: 0,
+			inputs: HashMap::new(),
+			prevtx_outpoints: HashSet::new(),
+			outputs: HashMap::new(),
+			feerate_sat_per_kw,
+		};
+		if is_initiator {
+			Self::RemoteChange(RemoteChange(context))
+		} else {
+			Self::LocalChange(LocalChange(context))
+		}
+	}
+
+	define_state_machine_transitions!(
+		LOCAL_OR_REMOTE_CHANGE, local_tx_add_input, remote_tx_add_input, &msgs::TxAddInput
+	);
+	define_state_machine_transitions!(
+		LOCAL_OR_REMOTE_CHANGE, local_tx_add_output, remote_tx_add_output, &msgs::TxAddOutput
+	);
+	define_state_machine_transitions!(
+		LOCAL_OR_REMOTE_CHANGE, local_tx_remove_input, remote_tx_remove_input, &msgs::TxRemoveInput
+	);
+	define_state_machine_transitions!(
+		LOCAL_OR_REMOTE_CHANGE, local_tx_remove_output, remote_tx_remove_output, &msgs::TxRemoveOutput
+	);
+	define_state_machine_transitions!(local_tx_complete, &msgs::TxComplete, [
+		FROM RemoteChange, TO LocalTxComplete,
+		FROM RemoteTxComplete, TO NegotiationComplete
+	]);
+	define_state_machine_transitions!(remote_tx_complete, &msgs::TxComplete, [
+		FROM LocalChange, TO RemoteTxComplete,
+		FROM LocalTxComplete, TO NegotiationComplete
+	]);
+}
+
+pub struct InteractiveTxConstructor<ES: Deref>
+	where
+		ES::Target: EntropySource,
+{
+	state_machine: StateMachine,
+	channel_id: [u8; 32],
+	is_initiator: bool,
+	entropy_source: ES,
+	inputs_to_contribute: Vec<(TxIn, Transaction)>,
+	outputs_to_contribute: Vec<TxOut>,
+}
+
+pub enum InteractiveTxMessageSend {
+	TxAddInput(msgs::TxAddInput),
+	TxAddOutput(msgs::TxAddOutput),
+	TxComplete(msgs::TxComplete),
+}
+
+macro_rules! do_state_transition {
+	($self: ident, $transition: ident, $msg: expr) => {{
+		let state_machine = core::mem::take(&mut $self.state_machine);
+		$self.state_machine = state_machine.$transition($msg);
+		match &$self.state_machine {
+			StateMachine::NegotiationAborted(_) => Err(()),
```

Comment:
Here it should return the inner error (from `NegotiationAborted`), instead of just ()